forch/ports_testing_states_manager.py

```diff
@@ -0,0 +1,111 @@
+"""Module that manages the testing states of the access ports"""
+
+import logging
+import threading
+
+from forch.proto.shared_constants_pb2 import TestingState
+
+
+LOGGER = logging.getLogger('portsm')
+STATE_HANDLERS = {}
+
+
+def _register_state_handler(state_name):
+    def register(func):
+        STATE_HANDLERS[state_name] = func
+        return func
+    return register
+
+
+class PortTestingStateMachine:
+    """State machine class that manages testing states of an access port"""
+
+    AUTHENTICATED = 'authenticated'
+    SEQUESTERED = 'sequestered'
+    OPERATIONAL = 'operational'
+    INFRACTED = 'infracted'
+
+    TRANSITIONS = {
+        AUTHENTICATED: {
+            TestingState.cleared: OPERATIONAL,
+            TestingState.sequestered: SEQUESTERED,
+        },
+        SEQUESTERED: {
+            TestingState.passed: OPERATIONAL,
+            TestingState.failed: INFRACTED,
+        },
+    }
+
+    def __init__(self, mac, initial_state):
+        self._mac = mac
+        self._current_state = initial_state
+
+    def handle_testing_state_event(self, testing_state):
+        """Handle testing state event"""
+        to_state = self.TRANSITIONS.get(self._current_state, {}).get(testing_state, {})
+
+        if not to_state:
+            LOGGER.warning(
+                'Cannot find next state for device %s in state %s for testing_state %s',
+                self._mac, self._current_state, testing_state)
+            return
+
+        LOGGER.info(
```

Comment:
Assuming a larger number of devices, would it be better to make this debug?

Reply:
I think we can firstly make it at INFO level just to make it easier to debug. After it has been properly running for some time, we can lower the severity to DEBUG...